### PR TITLE
refactor how output to stdout and stderr is written

### PR DESCRIPTION
**Pull request by soenkehahn**
This PR allows us to test all the output that is written to `stdout` and `stderr`, both by `check-protocol` and by its child processes.

This is in preparation for #87, but I think it could be reviewed independently.

### CODE REVIEW

**Comment by soenkehahn** [comment]
- [ ] Fix non-deterministic test errors

File: src/tracer/stdio_redirecting.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+impl Redirector {
+    pub fn new(context: &Context) -> R<Redirector> {
+        Ok(Redirector {
+            stdout: Redirect::new(context, StreamType::Stdout)?,
+            stderr: Redirect::new(context, StreamType::Stderr)?,
+        })
+    }
+
+    pub fn child(&self) -> R<()> {
```
Comment (matthandlersux) [inline src/tracer/stdio_redirecting.rs]: is there a more descriptive name for this other than "child"? similarly for "parent". i think i get what you're doing here, but the names don't give me much of a hint.

File: src/context.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
+use crate::R;
+use std::io::Write;
+use std::path::PathBuf;
+
+#[derive(Debug, Clone)]
+pub enum Context {
+    Context {
+        check_protocols_executable: PathBuf,
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply src/context.rs]: Just documentation. The `Context` differs between running the tool or running the tests. This field refers to the `check-protocols` executable. The path of that is needed for the mock executables.

File: src/context.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    use std::sync::{Arc, Mutex, MutexGuard};
+
+    #[derive(Debug, Clone)]
+    pub struct MockStream {
+        cursor: Arc<Mutex<Cursor<Vec<u8>>>>,
+    }
+
+    impl MockStream {
+        pub fn get_captured_stream(&self) -> String {
+            let cursor = self.cursor.lock().unwrap();
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply src/context.rs]: It's the `Deref` trait. Here's an explanation: https://doc.rust-lang.org/std/sync/struct.Arc.html#deref-behavior

File: src/context.rs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            Context::Context { .. } => Box::new(std::io::stdout()),
+            #[cfg(feature = "test")]
+            Context::TestContext { stdout, .. } => Box::new(stdout.clone()),
+        }
+    }
+
+    #[cfg(feature = "test")]
+    pub fn get_captured_stdout(&self) -> String {
+        match self {
+            Context::Context { .. } => panic!("tests should use the TestContext"),
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply src/context.rs]: @data-pup and I tried to do this with a trait. It's doable, but then e.g. `SyscallMock` and `Tracer` would need a type parameter for the concrete context type. And we thought that was rather ugly.

I'll try out doing this with conditional compilation though.

File: src/tracer/stdio_redirecting.rs (hunk 1 above)
Comment (soenkehahn) [reply src/tracer/stdio_redirecting.rs]: Done. Hope these names are better now.

**Pull request merged**